The Party and Sweets - CF 2P (#1349)
cpinitiative/usaco-guide: solutions/CF Party and Sweets.mdx

---
id: cf-1158A
source: CF
title: The Party and Sweets
author: Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/66993)

Let's consider the following example:

```
2 2
0 1
1 0
```

The answer to this example is $-1$ since boy $2$ ends up giving girl $2$ too
many sweets, even if boy $2$ gives girl $2$ the minimum number of sweets he
could possibly give, which is $1$. However, girl $2$ received a maximum of no
sweets, making it impossible.

If a single boy gives more sweets than a girl
received, then any arrangement of sweets is impossible with the constraints.
More formally, if

> $$max(b_1, b_2, \dots, b_n-1, b_n) > min(g_1, g_2, \dots, g_m-1, g_m)$$, then the

Reviewer: ```suggestion
$$max(b_1, b_2, \dots, b_{n-1}, b_n) > min(g_1, g_2, \dots, g_{m-1}, g_m)$$, then the
```

answer is $-1$.

If the input provided has an arrangement of sweets that follow the given
constraints, then we can greedily find the answer.

Since our objective is to minimize the total sum of sweets, it is optimal to
make each sweet in the arrangement as small as possible. Since the smallest
number of sweet a boy $i$ could give to a girl $j$ is $b_i$, we could "give"
each girl $b_i$ sweets, which is: $\sum\limits_{i=1}^{n} b_i \cdot m$.

However, with the constraints on the maximum number of sweets a girl $j$
received, we need to change $m$ of the sweet given to $g_j$. We greedily decide
on the sweet to replace (the $m$ greatest number of sweets given without
dissatisfying a constraint) by sorting $b_i$ and $g_j$ in decreasing order and
adding $\sum_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the
chosen $m$ sweets.

> sum.

Reviewer: ?

Author: ??

Reviewer: > sum.

Author: oh :man_facepalming:

### Implementation

**Time Complexity:** $\mathcal{O}(N\log N + M\log M)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int main(){
	int n, m; cin >> n >> m;
	vector<int> b(n), g(m);
	ll res = 0;

> 	for(int i=0; i<n; i++){

Reviewer: ```suggestion
	for(int i = 0; i < n; i++){
```

		cin >> b[i];
		res += b[i];
	}

> 	for(int i=0; i<m; i++){

Reviewer: ```suggestion
	for(int i = 0; i < m; i++){
```

		cin >> g[i];
	}
	// Sort b_i and g_i in descending order to improve the ease of using 2P
	sort(b.begin(), b.end(), [](int a, int b){
	   return a > b;
	});
	sort(g.begin(), g.end(), [](int a, int b){
	   return a > b;
	});
	// The initial sum without accounting for the constraints on the maximum values of g_i
	res *= m;
	int p1 = 0, p2 = 0, cur = 0;
	int maxB = b[0], minG = g[0];

> 	for(int i=1; i<n; i++){

Reviewer: ```suggestion
	for(int i = 1; i < n; i++){
```

		maxB = max(maxB, b[i]);
	}

> 	for(int i=1; i<m; i++){

Reviewer: ```suggestion
	for(int i = 1; i < m; i++){
```

		minG = min(minG, g[i]);
	}
	// If its impossible to satisfy the constraints
	if(maxB > minG){
		cout << -1 << endl;
		return 0;
	}
	// Two pointers to find the sum of (g_i - c_i).
	while(p2 < m){
		++cur;
		res += (g[p2] - b[p1]);

> 		if((cur == m-1 && g[p2+1] != b[p1]) || cur == m){

Reviewer: ```suggestion
		if((cur == m - 1 && g[p2 + 1] != b[p1]) || cur == m){
```

We are not done yet, however. The previous sum does not necessarily satisfy the condition
that $g_i$ is the maximum number of sweets a single girl received. Each girl $j$ has some boy $i$ who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know $\max(b) \leq \min(g)$, we would always choose the boy who gives the most sweets.

However, this boy still must give $b_i$ sweets to some girl; thus if no $g_j$ is equal to $b_i$, then he needs to have given some girl $b_i$ sweets and the boy who gives the second most sweets can give this girl her maximal sweets.

**Time Complexity:** $\mathcal{O}(N\log N + M\log M)$

<LanguageSection>

<CPPSection>

### C++ Implementation

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int main(){

	int n, m; cin >> n >> m;
	vector<int> b(n), g(m);
	ll res = 0;

	for(int i = 0; i < n; i++){
		cin >> b[i];
		res += b[i];
	}

	for(int i = 0; i < m; i++){
		cin >> g[i];
	}

	// Initial lower bound on answer
	res *= m;

> 	// Sort b_i and g_i in descending order to improve the ease of using 2P

Reviewer: ```suggestion
```
probably don't need this comment (since you're not sorting in descending order or using 2P)

	sort(b.begin(), b.end());
	sort(g.begin(), g.end());

	int maxB = b[0], minG = g[0];
	for(int i = 1; i < n; i++){
		maxB = max(maxB, b[i]);
	}
	for(int i = 1; i < m; i++){
		minG = min(minG, g[i]);
	}

	// If it's impossible to satisfy the constraints
	if(maxB > minG){
		cout << -1 << endl;
		return 0;
	}

	for(int i = 1; i < m; i++){ 

> 		res += g[i]-b[n-1];

Reviewer: ```suggestion
		res += g[i] - b[n - 1];
```

	}
	

> 	if(g[0] != b[n-1]){ 
> 		res += g[0]-b[n-2]; 

Reviewer: ```suggestion
	if(g[0] != b[n - 1]){ 
		res += g[0] - b[n - 2]; 
```

More formally, if
$$
max(b_1, b_2, \dots, b_{n-1}, b_n) > min(g_1, g_2, \dots, g_{m-1}, g_m),
$$ 
then the
answer is $-1$.

If the input provided has an arrangement of sweets that follow the given
constraints, then we can greedily find the answer.

Since our objective is to minimize the total sum of sweets, it is optimal to
make each sweet in the arrangement as small as possible. Since the smallest
number of sweet a boy $i$ could give to a girl $j$ is $b_i$, we could "give"
each girl $b_i$ sweets, which is: $\sum\limits_{i=1}^{n} b_i \cdot m$.

However, with the constraints on the maximum number of sweets a girl $j$
received, we need to change $m$ of the sweet given to $g_j$. We greedily decide
on the sweet to replace (the $m$ greatest number of sweets given without
dissatisfying a constraint) by sorting $b_i$ and $g_j$ in decreasing order and
adding $\sum_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the
chosen $m$ sweets.
sum.

### Implementation

**Time Complexity:** $\mathcal{O}(N\log N + M\log M)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int main(){

> 	int n, m; cin >> n >> m;

Reviewer 2: Maybe put each statement on their own line?

	vector<int> b(n), g(m);
	ll res = 0;
	for(int i = 0; i < n; i++){
		cin >> b[i];
		res += b[i];
	}
	for(int i = 0; i < m; i++){
		cin >> g[i];
	}
	// Sort b_i and g_i in descending order to improve the ease of using 2P
	sort(b.begin(), b.end(), [](int a, int b){
	   return a > b;
	});

> 	sort(g.begin(), g.end(), [](int a, int b){

Reviewer 2: I think you can do this with `std::greater<int>` or something like that, not sure though

	   return a > b;
	});
	// The initial sum without accounting for the constraints on the maximum values of g_i
	res *= m;
	int p1 = 0, p2 = 0, cur = 0;
	int maxB = b[0], minG = g[0];
	for(int i = 1; i < n; i++){
		maxB = max(maxB, b[i]);
	}

> 	for(int i = 1; i < m; i++){

Reviewer 2: Maybe add some spaces between those braces.

> 	int maxB = b[0], minG = g[0];

Reviewer 3: your choice: you could change this to
```suggestion
	int maxB = *max_element(begin(b), end(b));
	int minG = *min_element(begin(g), end(g));
```

		minG = min(minG, g[i]);
	}

> 	// If its impossible to satisfy the constraints

Reviewer 2: *it's

	if(maxB > minG){
		cout << -1 << endl;
		return 0;
	}

> 	int p1 = 0, p2 = 0, cur = 0;

Reviewer 3: ```suggestion
```
useless variables?

---
id: cf-1158A
source: CF
title: The Party and Sweets
author: Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/66993)

Let's consider the following example:

```
2 2
0 1
1 0
```

The answer to this example is $-1$ since boy $2$ ends up giving girl $2$ too
many sweets, even if boy $2$ gives girl $2$ the minimum number of sweets he
could possibly give, which is $1$. However, girl $2$ received a maximum of no
sweets, making it impossible.

If a single boy gives more sweets than a girl received, then any arrangement of
sweets is impossible with the constraints. More formally, if

$$

> max(b_1, b_2, \dots, b_{n-1}, b_n) > min(g_1, g_2, \dots, g_{m-1}, g_m),

Reviewer 3: ```suggestion
\max(b_1, b_2, \dots, b_{n-1}, b_n) > \min(g_1, g_2, \dots, g_{m-1}, g_m),
```

$$

then the answer is $-1$.

If the input provided has an arrangement of sweets that follow the given
constraints, then we can greedily find the answer.

Since our objective is to minimize the total sum of sweets, it is optimal to
make each sweet in the arrangement as small as possible. Since the smallest

> number of sweets a boy $i$ could give to a girl $j$ is $b_i$. Thus, the minimum number of sweets he could give out in total is $b_i * m$. Thus, the answer to this problem is lower bounded by $\sum\limits_{i=1}^{n} b_i \cdot m$.

Reviewer: ```suggestion
number of sweets a boy $i$ could give to a girl $j$ is $b_i$. Thus, the minimum number of sweets he could give out in total is $b_i \cdot m$. Thus, the answer to this problem is lower bounded by $\sum\limits_{i=1}^{n} b_i \cdot m$.
```

We are not done yet, however. The previous sum does not necessarily satisfy the condition

> that $g_i$ is the maximum number of sweets a single girl received. Each girl j has some boy i who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know max(b) <= min(g), we would always choose the boy who gives the most sweets.

Reviewer: ```suggestion
that $g_i$ is the maximum number of sweets a single girl received. Each girl j has some boy i who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know $\max(b) \leq \min(g)$, we would always choose the boy who gives the most sweets.
```
this should fix the error; i think the new parsing rules are stricter or something

> that $g_i$ is the maximum number of sweets a single girl received. Each girl j has some boy i who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know $\max(b) \leq \min(g)$, we would always choose the boy who gives the most sweets.

Reviewer 3: ```suggestion
that $g_i$ is the maximum number of sweets a single girl received. Each girl $j$ has some boy $i$ who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know $\max(b) \leq \min(g)$, we would always choose the boy who gives the most sweets.
```

**Time Complexity:** $\mathcal{O}(N\log N + M\log M)$

<LanguageSection>

<CPPSection>

### C++ Implementation

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int main(){

	int n, m; cin >> n >> m;
	vector<int> b(n), g(m);
	ll res = 0;

	for(int i = 0; i < n; i++){
		cin >> b[i];
		res += b[i];
	}

	for(int i = 0; i < m; i++){
		cin >> g[i];
	}

	// Initial lower bound on answer
	res *= m;

	// Sort b_i and g_i in descending order to improve the ease of using 2P
	sort(b.begin(), b.end(), greater<int>());
	sort(g.begin(), g.end(), greater<int>());

	int maxB = b[0], minG = g[0];
	for(int i = 1; i < n; i++){
		maxB = max(maxB, b[i]);
	}
	for(int i = 1; i < m; i++){
		minG = min(minG, g[i]);
	}

	// If it's impossible to satisfy the constraints
	if(maxB > minG){
		cout << -1 << endl;
		return 0;
	}

	int p1 = 0, p2 = 0, cur = 0;
	while(p2 < m){
		++cur;
		res += (g[p2] - b[p1]);
		if((cur == m - 1 && g[p2 + 1] != b[p1]) || cur == m){
			cur = 0;
			p1++;
		}
		p2++;

> 	}

Reviewer 4: Change to greedy (this 2-pointer is excessive).

somethign like:
```cpp
for(int i=1;i<m;++i) res += g[i]-b[n-1];
if(g[0] != b[n-1]) res += g[0]-b[n-2];
```

---
id: cf-1158A
source: CF
title: The Party and Sweets
author: Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/66993)

> **Time Complexity:** $\mathcal{O}(N\log N + M\log M)$

Reviewer: Move this under ## Implementation

The problem asks us to find the minimum sum of sweets which a boy could give to
a girl with the given constraints on the minimum number of sweets a boy could
give to a girl and the maximum number of sweets a girl has received from a boy.
Additionally, the problem asks us to print $-1$ if any given arrangement of

> sweets does not follow the given constraints.

Reviewer 5: why restate problem lol

Author: alr ill remove this lol

> Let's consider the following example from the sample input, which outputs $-1$:

Reviewer 5: ```suggestion
Let's consider the following example:
```

```
2 2
0 1
1 0
```

The answer to this example is $-1$ since boy $2$ ends up giving girl $2$ too
many sweets, even if boy $2$ gives girl $2$ the minimum number of sweets he
could possibly give, which is $1$. However, girl $2$ received a maximum of no
sweets, making it impossible.

> As we can see in this example, if a single boy gives more sweets than a girl

Reviewer 5: ```suggestion
If a single boy gives more sweets than a girl
```

received, then any arrangement of sweets is impossible with the constraints.

> More formally, if the

Reviewer: ```suggestion
More formally, if
```

> $max(b_1, b_2, \dots, b_n-1, b_n) > min(g_1, g_2, \dots, g_m-1, g_m)$, then the

Reviewer 5: Use $$ here

> Since our objective is to minimize the total sum of sweets, it is optimal to
> make each sweet in the arrangement as small as possible. Since the smallest
> number of sweets a boy $i$ could give to a girl $j$ is $b_i$. Thus, the minimum number of sweets he could give out in total is $b_i \cdot m$. Thus, the answer to this problem is lower bounded by $\sum\limits_{i=1}^{n} b_i \cdot m$.

Reviewer 4: This does not read well (I think I suggested it earlier, oops!)

```suggestion
Since our objective is to minimize the total number of sweets given out, let's first consider a lower bound on the answer. Since the fewest
number of sweets boy $i$ gives each girl is $b_i$, he gives out a total of at least $b_i \cdot m$ sweets. Thus, the answer to this problem is lower bounded by $\sum\limits_{i=1}^{n} b_i \cdot m$.
```

> If the input provided has an arrrangement of sweets that follows the given

Reviewer 5: ```suggestion
If the input provided has an arrangement of sweets that follow the given
```

Author: oops typo 🤦‍♂️

> constraints, then we can greedily search for the answer.

Reviewer 5: ```suggestion
constraints, then we can greedily find the answer.
```

Since our objective is to minimize the total sum of sweets, it is optimal to
make each sweet in the arrangement as small as possible. Since the smallest
number of sweet a boy $i$ could give to a girl $j$ is $b_i$, we could "give"

> each girl $b_i$ sweets, which is: $\sum_{i=1}^{n} b_i \times m$.

Reviewer: ```suggestion
each girl $b_i$ sweets, which is: $\sum\limits_{i=1}^{n} b_i \cdot m$.
```

> each girl $b_i$ sweets, which is: $\sum_{i=1}^{n} b_i \times m$.

Reviewer 5: delete everything before the equation and reword

Author: alright

However, with the constraints on the maximum number of sweets a girl $j$
received, we need to change $m$ of the sweet given to $g_j$. We greedily decide
on the sweet to replace (the $m$ greatest number of sweets given without
dissatisfying a constraint) by sorting $b_i$ and $g_j$ in decreasing order and

> add $\sum_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the

Reviewer: ```suggestion
add $\sum\limits_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the
```

> # Solution (Two Pointers + Greedy)

Reviewer 5: move above explanation

Reviewer 4: ^; Also: this is just Greedy (sorting is sorta implied/subset of greedy I think)

> add $\sum_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the

Reviewer 5: ```suggestion
adding $\sum_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the
```

<CPPSection>

> ### C++ Implementation

Reviewer 5: move outside

Reviewer: if you move it outside, change it to just ### Implementation

Reviewer 4: It's fine as is IMO

> chosen $m$ sweets. We can use the two-pointer technique to compute the previous

Reviewer 5: ```suggestion
chosen $m$ sweets.
```

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int main(){

	int n, m; cin >> n >> m;
	vector<int> b(n), g(m);
	ll res = 0;

	for(int i = 0; i < n; i++){
		cin >> b[i];
		res += b[i];
	}

	for(int i = 0; i < m; i++){
		cin >> g[i];
	}

> 	// The initial sum without accounting for the constraints on the maximum values of g_i

Reviewer: this line is too long

Reviewer 4: Perhaps change it to "Initial lower bound on answer" (because expl above clarifies)

---
id: cf-1158A
source: CF
title: The Party and Sweets
author: Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/66993)

Let's consider the following example:

```
2 2
0 1
1 0
```

The answer to this example is $-1$ since boy $2$ ends up giving girl $2$ too
many sweets, even if boy $2$ gives girl $2$ the minimum number of sweets he
could possibly give, which is $1$. However, girl $2$ received a maximum of no
sweets, making it impossible.

If a single boy gives more sweets than a girl received, then any arrangement of
sweets is impossible with the constraints. More formally, if

$$
max(b_1, b_2, \dots, b_{n-1}, b_n) > min(g_1, g_2, \dots, g_{m-1}, g_m),
$$

then the answer is $-1$.

If the input provided has an arrangement of sweets that follow the given
constraints, then we can greedily find the answer.

Since our objective is to minimize the total sum of sweets, it is optimal to
make each sweet in the arrangement as small as possible. Since the smallest

> number of sweet a boy $i$ could give to a girl $j$ is $b_i$, we could "give"
> each girl $b_i$ sweets, which is: $\sum\limits_{i=1}^{n} b_i \cdot m$.

Reviewer 4: ```suggestion
number of sweets a boy $i$ could give to a girl $j$ is $b_i$. Thus, the minimum number of sweets he could give out in total is $b_i * m$. Thus, the answer to this problem is lower bounded by $\sum\limits_{i=1}^{n} b_i \cdot m$.
```

> We are not done yet, however. The previous sum does not satisfy the condition

Reviewer 4: ```suggestion
We are not done yet, however. The previous sum does not necessarily satisfy the condition
```

that $g_i$ is the maximum number of sweets a single girl received. For this to
be satisfied, while still maintainting a minimum sum of sweets, we add
$\sum\limits_{i = 1}^{m} (g_i - c_i)$ to the answer, where $c_i$ represents the
chosen $m$ sweets to "replace" in order to satisfy $g_i$ being the maximum

> amount of candies a single girl received.

Reviewer 4: Probably something more along the lines of:

"
Each girl j has some boy i who gave her $g_j$ sweets instead of $b_i$ sweets. This raises our lower bound by $g_j-b_i$. We would like to choose $i$ such that $b_i$ is maximal. Since we know max(b) <= min(g), we would always choose the boy who gives the most sweets.

However, this boy still must give $b_i$ sweets to some girl; thus if no $g_j$ is equal to $b_i$, then he needs to have given some girl $b_i$ sweets and the boy who gives the second most sweets can give this girl her maximal sweets.
"

Also this doesn't use 2 pointer

Author: Yes it does. Look at my code.

Author: I basically moved the first pointer over if each $g_i$ is satisfied by replacing a given sweet with $g_i$ to $b_i$. I moved the second pointer over for each iteration. In essence, this is 2 pointers
